source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -1189,6 +1134,123 @@ bool ContextImpl::verifyCertChain(X509& leaf_cert, STACK_OF(X509) & intermediate
   return true;
 }
 
+void TlsContext::loadCertificateChain(const std::string& data, const std::string& data_path) {
+  cert_chain_file_path_ = data_path;
+  bssl::UniquePtr<BIO> bio(BIO_new_mem_buf(const_cast<char*>(data.data()), data.size()));
+  RELEASE_ASSERT(bio != nullptr, "");
+  cert_chain_.reset(PEM_read_bio_X509_AUX(bio.get(), nullptr, nullptr, nullptr));
+  if (cert_chain_ == nullptr || !SSL_CTX_use_certificate(ssl_ctx_.get(), cert_chain_.get())) {
+    logSslErrorChain();
+    throw EnvoyException(
+        absl::StrCat("Failed to load certificate chain from ", cert_chain_file_path_));
+  }
+  // Read rest of the certificate chain.
+  while (true) {
+    bssl::UniquePtr<X509> cert(PEM_read_bio_X509(bio.get(), nullptr, nullptr, nullptr));
+    if (cert == nullptr) {
+      break;
+    }
+    if (!SSL_CTX_add_extra_chain_cert(ssl_ctx_.get(), cert.get())) {
+      throw EnvoyException(
+          absl::StrCat("Failed to load certificate chain from ", cert_chain_file_path_));
+    }
+    // SSL_CTX_add_extra_chain_cert() takes ownership.
+    cert.release();
+  }
+  // Check for EOF.
+  const uint32_t err = ERR_peek_last_error();
+  if (ERR_GET_LIB(err) == ERR_LIB_PEM && ERR_GET_REASON(err) == PEM_R_NO_START_LINE) {
+    ERR_clear_error();
+  } else {
+    throw EnvoyException(
+        absl::StrCat("Failed to load certificate chain from ", cert_chain_file_path_));
+  }
+}
+
+void TlsContext::loadPrivateKey(const std::string& data, const std::string& data_path,
+                                const std::string& password) {
+  bssl::UniquePtr<BIO> bio(BIO_new_mem_buf(const_cast<char*>(data.data()), data.size()));
+  RELEASE_ASSERT(bio != nullptr, "");
+  bssl::UniquePtr<EVP_PKEY> pkey(
+      PEM_read_bio_PrivateKey(bio.get(), nullptr, nullptr,
+                              !password.empty() ? const_cast<char*>(password.c_str()) : nullptr));
+
+  if (pkey == nullptr || !SSL_CTX_use_PrivateKey(ssl_ctx_.get(), pkey.get())) {
+    throw EnvoyException(fmt::format("Failed to load private key from {}, Cause: {}", data_path,
+                                     Utility::getLastCryptoError().value_or("unknown")));
+  }
+
+  checkPrivateKey(pkey, data_path);
+}
+
+void TlsContext::loadPkcs12(const std::string& data, const std::string& data_path,
+                            const std::string& password) {
+  cert_chain_file_path_ = data_path;
+  bssl::UniquePtr<BIO> bio(BIO_new_mem_buf(const_cast<char*>(data.data()), data.size()));
+  RELEASE_ASSERT(bio != nullptr, "");
+  bssl::UniquePtr<PKCS12> pkcs12(d2i_PKCS12_bio(bio.get(), nullptr));
+
+  EVP_PKEY* temp_private_key = nullptr;
+  X509* temp_cert = nullptr;
+  STACK_OF(X509)* ca_certificates = nullptr;
```

Comment:
Can this be a `bssl::UniquePtr<STACK_OF(X509_INFO)>`?

Reply:
As far as I can tell, it can't be for two reasons:
1) the macro magic around `STACK_OF()` confuses the template engine and it can't figure out the deleter to use.
2) the address-of operator is not permitted on the `bssl::UniquePtr` type, and this is needed as an out parameter from `PKCS12_parse`.

Reply:
> the macro magic around STACK_OF() confuses the template engine and it can't figure out the deleter to use.

There are other uses of this type, but they all have an immediate initialization, eg `bssl::UniquePtr<STACK_OF(X509_INFO)> list(PEM_X509_INFO_read_bio(bio.get(), nullptr, nullptr, nullptr));` 

> he address-of operator is not permitted

Got it.

Can you create a bssl::UniquePtr to own the memory on the next line, right next to the other two? It's easier to reason about the function when all the memory is handled in the same way (RAII ownership).